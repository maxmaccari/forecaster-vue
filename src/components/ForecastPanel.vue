<template>
  <div>Forecast Here</div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  async setup() {
    await new Promise((resolve, reject) => {
      setTimeout(() => reject({ message: 'my error' }), 1000)
    })

    return {}
  },
})
</script>
