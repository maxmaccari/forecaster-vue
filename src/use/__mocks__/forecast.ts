export const useForecast = jest.fn()