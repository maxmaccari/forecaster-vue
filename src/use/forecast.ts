declare interface WindResponse {
  speed: number
  deg: number
}

declare interface WeatherResponse {
  coord: { lon: number; lat: number }
  weather: {
    id: number
    main: string
    description: string
    icon: string
  }[]
  base: string
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    humidity: number
  }
  visibility: number
  wind: WindResponse
  clouds: {
    all: number
  }
  dt: number
  sys: {
    type: number
    id: number
    country: string
    sunrise: number
    sunset: number
  }
  timezone: number
  id: number
  name: string
  cod: number
}

declare interface ListResponse {
  dt: number
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    sea_level: number
    grnd_level: number
    humidity: number
    temp_kf: number
  }
  weather: [
    {
      id: number
      main: string
      description: string
      icon: string
    }
  ]
  clouds: { all: number }
  wind: WindResponse
  visibility: number
  pop: number
  sys: { pod: string }
  dt_txt: string
}

declare interface ForecastResponse {
  cod: string
  message: number
  cnt: number
  list: ListResponse[]
  city: {
    id: number
    name: string
    coord: { lat: number; lon: number }
    country: string
    population: number
    timezone: number
    sunrise: number
    sunset: number
  }
}

export class Weather {
  description: string | null = null
  icon: string | null = null
  temperature = 0
  minTemperature = 0
  maxTemperature = 0
  feelsLike = 0
  humidity = 0
  pressure = 0

  constructor(weather: WeatherResponse | ListResponse | null = null) {
    if (!weather) return

    this.description = weather.weather[0]
      ? weather.weather[0].description
      : null
    this.icon = weather.weather[0] ? weather.weather[0].icon : null
    this.temperature = weather.main.temp
    this.minTemperature = weather.main.temp_min
    this.maxTemperature = weather.main.temp_max
    this.feelsLike = weather.main.feels_like
    this.humidity = weather.main.humidity
    this.pressure = weather.main.pressure
  }
}

export class Wind {
  direction = 0
  speed = 0

  constructor(windResponse: WindResponse | null = null) {
    this.direction = windResponse?.deg || 0
    this.speed = windResponse?.speed || 0
  }
}

export class ForecastDetails {
  clouds = 0
  timestamp = 0
  weather: Weather = new Weather()
  wind: Wind = new Wind()

  constructor(details: ListResponse) {
    this.clouds = details.clouds.all
    this.timestamp = details.dt
    this.weather = new Weather(details)
  }
}

export class Forecast {
  name = ''
  country = ''
  weather: Weather = new Weather()
  wind: Wind = new Wind()
  details: ForecastDetails[] = []

  constructor(weather: WeatherResponse, forecast: ForecastResponse) {
    this.name = weather.name
    this.country = forecast.city.country
    this.weather = new Weather(weather)
    this.wind = new Wind(weather.wind)
    this.details = forecast.list.map(details => {
      return new ForecastDetails(details)
    })
  }
}

class CacheHash {
  [Identifier: string]: Forecast
}

let cache: CacheHash = new CacheHash()

const API_URL = process.env.VUE_APP_API_URL
const API_KEY = process.env.VUE_APP_API_KEY

export const fetchWeather = (location: string): Promise<WeatherResponse> => {
  const url = `${API_URL}/weather?q=${location}&appid=${API_KEY}&units=metric`

  return window.fetch(url).then(response => {
    return (response.json() as Record<string, any>) as WeatherResponse
  })
}

export const fetchForecast = (location: string): Promise<ForecastResponse> => {
  const url = `${API_URL}/forecast?q=${location}&appid=${API_KEY}&units=metric`

  return window.fetch(url).then(response => {
    return (response.json() as Record<string, any>) as ForecastResponse
  })
}

export const clearCache = () => {
  cache = new CacheHash()
}

export const useForecast = async (location: string): Promise<Forecast> => {
  if (cache[location]) return cache[location]

  const [weatherJson, forecastJson] = await Promise.all([
    fetchWeather(location),
    fetchForecast(location),
  ])

  const forecast = new Forecast(weatherJson, forecastJson)
  cache[location] = forecast

  return forecast
}
