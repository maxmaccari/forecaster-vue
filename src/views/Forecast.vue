<template>
  <div class="flex items-center justify-center min-h-screen">
    <AsyncBoundary>
      <AsyncProvider :provider="provider" v-slot="{ data }">
        <ForecastPanel :forecast="data" />
      </AsyncProvider>
    </AsyncBoundary>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import ForecastPanel from '@/components/ForecastPanel.vue'
import AsyncBoundary from '@/components/AsyncBoundary.vue'
import AsyncProvider from '@/components/AsyncProvider.vue'

export default defineComponent({
  components: { ForecastPanel, AsyncBoundary, AsyncProvider },
  props: {
    location: {
      type: String,
      required: true,
    },
  },
  methods: {
    async provider() {
      return new Promise(resolve => {
        setTimeout(() => resolve({ message: 'hello world' }), 1000)
      })
    },
  },
})
</script>
