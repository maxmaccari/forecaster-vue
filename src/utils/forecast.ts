import { Forecast } from '@/use/forecast'

export class NextHoursSummary {
  constructor(
    public temperature: number,
    public time: string,
    public clouds: number,
    public icon: string | null
  ) {}
}

export class NextWeekEntry {
  constructor(
    public day: string,
    public clouds: number,
    public min: number,
    public temperature: number,
    public icon: string | null,
    public description: string | null
  ) {}
}

export const dateFormat = new Intl.DateTimeFormat('en-US', {
  day: 'numeric',
  weekday: 'long',
})

export const timeFormat = new Intl.DateTimeFormat('en-US', {
  hour: 'numeric',
  minute: 'numeric',
})

export const getNextHours = (forecast: Forecast): NextHoursSummary[] => {
  return [forecast.details[0], forecast.details[2], forecast.details[4]].map(
    detail => new NextHoursSummary(
      Math.round(detail.weather.temperature),
      timeFormat.format(new Date(detail.timestamp * 1000)),
      detail.clouds,
      detail.weather.icon,
    )
  )
}

export const getNextWeek = (forecast: Forecast): NextWeekEntry[] => {
  return forecast.details.reduce((week: NextWeekEntry[], detail) => {
    const date = new Date(detail.timestamp * 1000)
    const day = dateFormat.format(date)
    const forecast = week.find(forecast => forecast.day === day)

    if (!forecast) {
      week.push(new NextWeekEntry(day, 
        detail.clouds,
        Math.round(detail.weather.minTemperature),
        Math.round(detail.weather.temperature),
        detail.weather.icon,
        detail.weather.description,
      ))
    } else {
      if (detail.weather.minTemperature < forecast.min) {
        forecast.min = Math.round(detail.weather.minTemperature)
      }

      if (detail.weather.temperature > forecast.temperature) {
        forecast.temperature = Math.round(detail.weather.temperature)
      }

      if (date.getHours() > 10 && date.getHours() < 14) {
        forecast.icon = detail.weather.icon
      }
    }

    return week
  }, [])
}
