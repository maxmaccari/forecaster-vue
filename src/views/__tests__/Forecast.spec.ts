import { describe, it, expect, vi, MockedFunction } from 'vitest';

import { shallowMount, mount, flushPromises } from '@vue/test-utils'
import Forecast from '../Forecast.vue'
import ForecastPanel from '@/components/ForecastPanel.vue'
import { useForecast } from '@/use/forecast'
import { buildForecast } from '@unit/helpers/forecast'

vi.mock('@/use/forecast')
const mockedUseForecast = useForecast as MockedFunction<typeof useForecast>

describe('Forecast', () => {
  it('renders the view properly', () => {
    const location = 'barcelona'
    const wrapper = shallowMount(Forecast, { props: { location } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders ForecastPanel passing the location and the loaded forecast data as props', async () => {
    const data = buildForecast()
    mockedUseForecast.mockResolvedValue(data)
    const location = 'barcelona'
    const wrapper = mount(Forecast, {
      props: { location },
      global: {
        stubs: ['forecast-panel'],
      },
    })

    await flushPromises()
    await wrapper.vm.$nextTick()

    const forecastPanelWrapper = wrapper.findComponent(ForecastPanel)

    expect(forecastPanelWrapper.exists()).toBe(true)
    expect(forecastPanelWrapper.props('location')).toBe(location)
    expect(forecastPanelWrapper.props('forecast')).toEqual(data)
  })
})
