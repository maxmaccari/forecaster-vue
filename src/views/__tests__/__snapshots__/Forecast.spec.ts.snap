// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Forecast > renders the view properly 1`] = `
"<div class="flex items-center justify-center min-h-screen">
  <async-boundary-stub></async-boundary-stub>
</div>"
`;
